File: src/venture_guide/lib/app/map/domain/services/tile_service.dart
import 'dart:math';
import 'package:injectable/injectable.dart';
import 'package:latlong2/latlong.dart';

@Injectable()
class TitleService {
  final double zoom = 7;

  Point<int> latLngToTile(LatLng latLng) {
    double lat = latLng.latitude;
    double lng = latLng.longitude;

    // Converter coordenadas para radianos
    double latRad = lat * pi / 180;

    // Cálculo da projeção de Mercator
    double mercN = log(tan((pi / 4) + (latRad / 2)));

    // Calculando as coordenadas do tile
    int x = ((lng + 180) / 360 * pow(2, zoom)).toInt();
    int y = ((1 - mercN / pi) / 2 * pow(2, zoom)).toInt();

    return Point<int>(x, y);
  }

  List<Point<int>> latLngBoundsToTiles(LatLng latLng1, LatLng latLng2) {
    final southWest = latLngToTile(latLng1);
    final northEast = latLngToTile(latLng2);

    var tiles = <Point<int>>[];
    for (var lat = northEast.y; lat <= southWest.y; lat++) {
      for (var lon = southWest.x; lon <= northEast.x; lon++) {
        tiles.add(Point(lon, lat));
      }
    }
    return tiles;
  }
}

File: src/venture-guide-firebase/functions/src/services/tileService.ts
export interface LatLng {
    latitude: number;
    longitude: number;
}

export interface Point {
    x: number;
    y: number;
}

const zoom = 7;
export const latLngToTile = (latLng: LatLng): Point => {
    const lat = latLng.latitude;
    const lng = latLng.longitude;

    // Converter coordenadas para radianos
    const latRad = lat * Math.PI / 180;

    // Cálculo da projeção de Mercator
    const mercN = Math.log(Math.tan((Math.PI / 4) + (latRad / 2)));

    // Calculando as coordenadas do tile
    const x = Math.floor((lng + 180) / 360 * Math.pow(2, zoom));
    const y = Math.floor((1 - mercN / Math.PI) / 2 * Math.pow(2, zoom));

    return { x, y };
}
